src/commonMain/kotlin/fr/acinq/lightning/serialization/v4/Serialization.kt

```diff
@@ -0,0 +1,465 @@
+        writeLightningMessage(remoteShutdown)
+        writeNullable(closingFeerates) { writeClosingFeerates(it) }
+    }
+
+    private fun Output.writeNegotiating(o: Negotiating) = o.run {
+        writeCommitments(commitments)
+        writeLightningMessage(localShutdown)
+        writeLightningMessage(remoteShutdown)
+        writeCollection(closingTxProposed) {
+            writeCollection(it) { closingTxProposed ->
+                closingTxProposed.run {
+                    writeTransactionWithInputInfo(unsignedTx)
+                    writeLightningMessage(localClosingSigned)
+                }
+            }
+        }
+        writeNullable(bestUnpublishedClosingTx) { writeTransactionWithInputInfo(it) }
+        writeNullable(closingFeerates) { writeClosingFeerates(it) }
+    }
+
+    private fun Output.writeClosing(o: Closing) = o.run {
+        writeCommitments(commitments)
+        writeNullable(fundingTx) { writeBtcObject(it) }
+        writeNumber(waitingSinceBlock)
+        writeCollection(alternativeCommitments) { writeCommitments(it) }
+        writeCollection(mutualCloseProposed) { writeTransactionWithInputInfo(it) }
+        writeCollection(mutualClosePublished) { writeTransactionWithInputInfo(it) }
+        writeNullable(localCommitPublished) { writeLocalCommitPublished(it) }
+        writeNullable(remoteCommitPublished) { writeRemoteCommitPublished(it) }
+        writeNullable(nextRemoteCommitPublished) { writeRemoteCommitPublished(it) }
+        writeNullable(futureRemoteCommitPublished) { writeRemoteCommitPublished(it) }
+        writeCollection(revokedCommitPublished) { writeRevokedCommitPublished(it) }
+    }
+
+    private fun Output.writeLocalCommitPublished(o: LocalCommitPublished) = o.run {
+        writeBtcObject(commitTx)
+        writeNullable(claimMainDelayedOutputTx) { writeTransactionWithInputInfo(it) }
+        writeCollection(htlcTxs.entries) {
+            writeBtcObject(it.key)
+            writeNullable(it.value) { htlcTx -> writeTransactionWithInputInfo(htlcTx) }
+        }
+        writeCollection(claimHtlcDelayedTxs) { writeTransactionWithInputInfo(it) }
+        writeCollection(claimAnchorTxs) { writeTransactionWithInputInfo(it) }
+        writeIrrevocablySpent(irrevocablySpent)
+    }
+
+    private fun Output.writeRemoteCommitPublished(o: RemoteCommitPublished) = o.run {
+        writeBtcObject(commitTx)
+        writeNullable(claimMainOutputTx) { writeTransactionWithInputInfo(it) }
+        writeCollection(claimHtlcTxs.entries) {
+            writeBtcObject(it.key)
+            writeNullable(it.value) { claimHtlcTx -> writeTransactionWithInputInfo(claimHtlcTx) }
+        }
+        writeCollection(claimAnchorTxs) { writeTransactionWithInputInfo(it) }
+        writeIrrevocablySpent(irrevocablySpent)
+    }
+
+    private fun Output.writeRevokedCommitPublished(o: RevokedCommitPublished) = o.run {
+        writeBtcObject(commitTx)
+        write(remotePerCommitmentSecret.value.toByteArray())
```

Comment:
nit: every time we use `Output.write()` directly  and not `writeDelimited()` it's because the value being serialized has a fixed, known size (32, 33, 64 bytes ...), why not add explicit methods for serializing/deserializing  Bytevector32/Bytevector64/etc... as using `Output.write()` directly is potentially dangerous ?

Reply:
Done in [bb24e14](https://github.com/ACINQ/lightning-kmp/pull/391/commits/bb24e14ae59fe4d021bd2749802d721aaa111bdc).